### PR TITLE
Store QAQC in Digital Ocean instead of Publishing Database

### PR DESCRIPTION
**Pull request by SashaWeinstein**
# Update storage of PLUTO QAQC Tables
This PR is to address issue #317 
Before you make any comments be sure to note that **this PR is not done yet**. Currently this PR is here to show the team  the work I've done thus far. There is a list of outstanding questions at the bottom, and once those are resolved then this can treated like a typical PR and eventually approved and merged. 

## How does PLUTO QAQC work?
The QAQC on this data product is different from our other data products in that new results are *added as rows to existing tables*. So the QAQC scripts need access to the results of the previous QAQC, which makes the process more complex.
Currently the results of QAQC on previous runs are all stored in the publishing database (the path to this DB is assigned to `EDM_DATA` in the `.env` files), and new results appended to these tables. 

## Why do we need an upgrade?
Currently the QAQC process writes directly to those 4 tables on the publishing database regardless of where it was run. This makes the info displayed on the [QAQC app](https://edm-data-engineering.nycplanningdigital.com/?page=Home) less reliable as QAQC on a testing run on someone's local machine could be displayed. Another issue is that one can't compares QAQC results run off different branches like one can for facDB. And making changes/additions to the QAQC scripts is extra hard as one has to be careful about messing with this important source of historical data. 

## New approach 
The new approach follows facDB's process of [writing QAQC tables to .csv's and putting them in an output folder that gets written to DO](https://github.com/NYCPlanning/db-facilities/blob/13447e36ce214bbc907c56b6f9d4740ef551e232/facdb/bash/export.sh#L11). This repo's implementation is more complex as the existing QAQC files must be pulled down first, and tables exported as .csv's for the app and .sql files for future builds. 

### Dataloading
Based this work off the [similar process in facDB](https://github.com/NYCPlanning/db-facilities/blob/13447e36ce214bbc907c56b6f9d4740ef551e232/facdb/bash/dataloading.sh#L6). One additional thing is to rename table with a SQL query written into the bash script

### QAQC Step
This was under the archive step, but I moved it to it's own bash script as I don't think QAQC and archive are very similar. 
I also refactored the code to loop through files and combinations of `CONDO` and `MAPPED` programmatically which made the code shorter and more readable. Previously the `CONDITION` was hardcoded and could be changed independently of the `CONDO` and `MAPPED` options which is quite fragile, but now it's set programmatically. 

### QAQC .sql files
Changes to these files is simple, point to tables in the local database and cast bbl's to float as types can get changed in the dataloading process

### Import qaqc function
This is based off the `import_public` function. We need this as the PLUTO QAQC process requires having QAQC on all versions/pairs in the same table

## Outstanding questions
- Is the `archive_pluto` the best table to use as "current"? There is also `export_pluto` and `pluto`
- Where should the qaqc .sql files be pulled from? Currently `05_qaqc.sql` points to the 317-QAQC-to_DO folder in `edm-publishing/db-pluto` but this is just for testing. I think that reading from the `latest` folder makes the most sense. This begs a larger question of whether all runs should upload files to the `latest` folder. We could implement something like facDB where files are uploaded to a folder named after the branch to avoid the common problem of test runs writing to the `latest` folder. This is preferable as it can be assumed that the output under `main` is correct.
- Where in the code should import_qaqc be called? I put in in `05_qaqc.sh` but it could live in the dataloading step as well. And where should it store the `.sql` files? I made a new sub-folder under `.library` but this is misleading as the data doesn't come from data library at all. It's also slightly confusing that the files are called `qaqc_expected.sql` when the code to *generate* that QAQC table has the same name. Curious if others find this confusing as well or if it's good enough

## How to test
You should be able to test without re-building. First run the new code in the dataloading step, either by commenting everything else out or running it from the command line. Once you have the previous pluto, you should be able to run the QAQC bash script on its own. Come to me of course with any questions on how to test this code locally.

### CODE REVIEW

**Comment by abrieff** [comment]
> Where in the code should import_qaqc be called? I put in in 05_qaqc.sh but it could live in the dataloading step as well. And where should it store the .sql files? I made a new sub-folder under .library but this is misleading as the data doesn't come from data library at all. It's also slightly confusing that the files are called qaqc_expected.sql when the code to generate that QAQC table has the same name. Curious if others find this confusing as well or if it's good enough

1. I like the separate 05_qaqc.sh step - it's already such a long process to dataload and build, and I think it's a reasonably different set of steps. 
2. I think its reasonable that .library means the local library of downloaded files, not the remote data-library specifically.
3. Could the code to generate QAQC table be create_qaqc_expected.sql or something along those lines to differentiate?

**Comment by Oysters1874** [comment]
I agree with Andrew on the point of separating 05_qaqc.sh, as it is another component somehow and it makes more sense to do it later than the real dataloading step. The .library thing didn't confused me too much, it seems convenient to put them all together. Also, for the question of uploading all runs to the latest folder, I agree with you on using another approach to store all test runs to aviod messing things up.

**Comment by SashaWeinstein** [comment]
Thanks for the feedback!

**Review by abrieff — COMMENTED** [review]
This is failing for me locally because these qaqc tables don't already exist. Where should the create table line be running?

**Review by abrieff — APPROVED** [review]

**Comment by abrieff** [comment]
Some of the commands included in the qaqc files imported from 'import_qaqc' include lines that are incompatible with Postgres v11, which seems to be the canonical version of postgres for pluto according to Digital Ocean. 

For example,
```
SET default_table_access_method = heap;
```
fails on v11.

We either need to establish a new upgraded version for pluto (risky but we're a few versions back in general, this might be worth doing at some point) or make sure that any of these sql files are v11 compatible. 

Line from top of QAQC_Expected:

```
--
-- PostgreSQL database dump
--

-- Dumped from database version 13.4 (Debian 13.4-4.pgdg110+1)
-- Dumped by pg_dump version 14.0
```
This is probably the issue

**Comment by SashaWeinstein** [comment]
Ok word. Can you try running not in the devcontainer? Then we can make sure it works as expected with the newer postgres, and then we can worry about the postgres version used in PLUTO

File: pluto_build/05_qaqc.sh
Comment (SashaWeinstein) [reply pluto_build/05_qaqc.sh:39]: making this change now!

**Comment by abrieff** [comment]
I can, but i'm not sure that matters if we're not upgrading pluto's postgres more broadly? I think that you probably need to redump the qaqc files to DO using v11 (if I understand correctly that what's being pulled down is sqldumps you pushed up at the beginning of this process)

**Comment by SashaWeinstein** [comment]
I think it does matter, as we want to make sure the logic works. Making sure the process works and getting the database versions to match should be treated as two different issues

**Comment by abrieff** [comment]
Valid - rerunning using v12 - will take awhile to build. It looks like github actions is using v12 which is probably partially why this wouldn't have raised an error - would still be concerned about this as a roadblock before merging/approving though

**Comment by SashaWeinstein** [comment]
I definitely agree it's an issue to be resolved before merging

**Comment by SashaWeinstein** [comment]
Ok so I looked at the postgres versions across the publishing database, stand-alone DB container, and DB container in the docker compose devcontainer setup. 
- Publishing DB has postgres 11.6
- Stand-alone DB container is on version 13.4 (this is what I used to test the new QAQC process)
- DB container in docker compose stack has version 11.1

I think upgrading the DB container in docker compose is as easy as pulling from a newer postgis image in the docker compose .yml. Upgrading the publishing DB is probably harder but I don't know if that's necessary for this improvement as the .sql files being download/uploaded are separate now. 

Andrew what did you mean when you said "It looks like github actions is using v12"? Is that the version of postgres of the DB container created during a github hosted action (pre-.devcontainer implementation of course)

**Comment by abrieff** [comment]
> Andrew what did you mean when you said "It looks like github actions is using v12"? Is that the version of postgres of the DB container created during a github hosted action (pre-.devcontainer implementation of course)

Yes - this line in github actions pre-devcontainer
    services:
      postgres:
        image: postgis/postgis:12-3.0-alpine

**Comment by SashaWeinstein** [comment]
Do you need to update the version of pgdump you're using?

File: pluto_build/01_dataloading.sh
Comment (td928) [inline pluto_build/01_dataloading.sh:56]: might be a good cleanup to remove space

**Comment by mbh329** [comment]
1. Concur with everyone else that the 05_qaqc.sh step is a nice addition to break up the archive and qaqc step which as you mentioned are two distinct steps
2. Glad we settled on that today at standup - is there any value in cleaning up the digital ocean folders every once and a while? Or maybe limiting the qaqc pages to only have more recent branches for comparison? Looking at facdb and pluto there are many branch folders that haven't been touched in years. I don't think storage is much of an issue for us currently on DO so not sure if any of these comments really matter
3. Will look further into this

File: pluto_build/bin/config.sh
Comment (td928) [inline pluto_build/bin/config.sh:97]: I see this now follows the convention we discussed in standup today. Looks good to me!

**Comment by SashaWeinstein** [comment]
@mbh329 thanks for volunteering to clean up the facDB branches listed on the app! I've scoped out the issue [here](https://github.com/NYCPlanning/data-engineering-qaqc/issues/51), it should only take a couple lines of logic to remove the outdated branches

File: pluto_build/bin/config.sh
Comment (mbh329) [inline pluto_build/bin/config.sh:160]: I am getting a `bin/config.sh: line 160: register: command not found` for anytime the register command is called in the config.sh - it doesn't seem to impact the dataloading but was curious if anyone else saw this as well? Is this a relic of the ancient version of postgres yall were discussing?

Comment (SashaWeinstein) [reply pluto_build/bin/config.sh:160]: I've seen that error message since I started working here and always ignored it. I think it's from broken bash code, not postgres

File: pluto_build/bin/config.sh
Comment (SashaWeinstein) [reply pluto_build/bin/config.sh:97]: te can I get an approval of the PR itself so we can merge in?

Comment (td928) [reply pluto_build/bin/config.sh:97]: Hold on still trying to test everything. Will give you a heads up when I finished. Thanks!

File: pluto_build/bin/config.sh
Comment (Oysters1874) [reply pluto_build/bin/config.sh:160]: I am getting it every time and it seems like nothing is impacted, so I just ignore.

File: pluto_build/bin/config.sh
Comment (SashaWeinstein) [reply pluto_build/bin/config.sh:97]: ah ok great, sorry didn't realize you were still working

**Comment by mbh329** [comment]
3. I don't love keeping the .sql files from the qaqc.sh in the .library folder but I think its good enough

**Review by td928 — APPROVED** [review]
sorry taking so long but local build took me awhile with the memory issue earlier. Everything looks good to me now! Thanks!

**Pull request merged**
